cmake/compiler_support.cmake

```diff
@@ -0,0 +1,261 @@
+   set(${COMPILER_NAME}_DEFINES ${COMPILER_DEFINES}       CACHE INTERNAL "Defines for compiler ${COMPILER_NAME}") 
+   set(${COMPILER_NAME}_INCLUDES 
+      ${${COMPILER_NAME}_ROOT}/${TR_TARGET_ARCH}/${TR_TARGET_SUBARCH}
+      ${${COMPILER_NAME}_ROOT}/${TR_TARGET_ARCH} 
+      ${${COMPILER_NAME}_ROOT} 
+      ${OMR_ROOT}/compiler/${TR_TARGET_ARCH}/${TR_TARGET_SUBARCH}
+      ${OMR_ROOT}/compiler/${TR_TARGET_ARCH} 
+      ${OMR_ROOT}/compiler 
+      ${OMR_ROOT}
+      ${COMPILER_INCLUDES}
+      CACHE INTERNAL "Include header list for ${COMPILER}")
+
+   message("${COMPILER_NAME}_ROOT = ${${COMPILER_NAME}_ROOT}") 
+
+
+   # Generate a build name file. 
+   set(BUILD_NAME_FILE "${CMAKE_BINARY_DIR}/${COMPILER_NAME}Name.cpp")
+   add_custom_command(OUTPUT ${BUILD_NAME_FILE}  
+                     COMMAND perl ${OMR_ROOT}/tools/compiler/scripts/generateVersion.pl ${COMPILER_NAME} > ${BUILD_NAME_FILE}
+		     VERBATIM
+                     BYPRODUCTS ${BUILD_NAME_FILE}
+                     COMMENT "Generate ${BUILD_NAME_FILE}"
+                     )
+
+   # Run masm2gas on contained .asm files 
+   masm2gas_asm_files(COMPILER_OBJECTS ${COMPILER_NAME} ${COMPILER_OBJECTS})
+
+   add_library(${COMPILER_NAME} ${LIB_TYPE} 
+      ${BUILD_NAME_FILE} 
+      ${COMPILER_OBJECTS}
+      )  
+
+   # Populated and then will use target_sources to append to the compiler library.
+   set(CORE_COMPILER_OBJECTS "")
+
+   # Add the contents of the macro to CORE_COMPILER_OBJECTS in this scope.  
+   # The library name parameter is currently ignored.
+   macro(compiler_library libraryname)
+      set(CORE_COMPILER_OBJECTS ${CORE_COMPILER_OBJECTS} ${ARGN} PARENT_SCOPE)
+   endmacro(compiler_library)
+
+   # Because we want to be able to build multiple compiler 
+   # components in the same tree (ie the compiler test code, as 
+   # well as JitBuilder, we can't just use the plain add_subdirectory, 
+   # CMake demands we specify the object target directory as well. 
+   # 
+   # This macro computes an appropriate target directory.
+   macro(add_compiler_subdirectory dir) 
+      add_subdirectory(${OMR_ROOT}/compiler/${dir}
+                       ${CMAKE_CURRENT_BINARY_DIR}/compiler/${dir}_${COMPILER_NAME})
+   endmacro(add_compiler_subdirectory)
+
+
+   # There's a little bit of a song and dance here 
+   # I wonder if it's not better to just have a 
+   # unified object list. 
+   add_compiler_subdirectory(ras)
+   add_compiler_subdirectory(compile)
+   add_compiler_subdirectory(codegen)
+   add_compiler_subdirectory(control)
+   add_compiler_subdirectory(env)
+   add_compiler_subdirectory(infra)
+   add_compiler_subdirectory(il)
+   add_compiler_subdirectory(optimizer)
+   add_compiler_subdirectory(runtime)
+   add_compiler_subdirectory(ilgen)
+
+   if(${TR_TARGET_ARCH} STREQUAL "x")
+      add_compiler_subdirectory(x)
+   endif()
+
+
+   # Filter out objects requested to be removed by 
+   # client project (if any): 
+   foreach(object ${COMPILER_FILTER})
```

Comment:
/me nods. There's a longer term question of defining what is _core_ to the compiler, and cannot be overridden. However, in the short term, I think it's how we need to proceed to allow client projects to customize until we can lay down that particular definition.